### PR TITLE
kvclient,server: propagate pprof labels for BatchRequests

### LINKED ISSUE #100166: kvserver: propagate job pprof labels to KV requests sent on behalf of the job
Similar to SQL queries (https://github.com/cockroachdb/cockroach/issues/86012) jobs would also like their pprof labels to be propagated to kvserver so that requests issued on behalf of the job are tagged as such when profiling is enabled. We already propagate pprof labels across DistSQL boundaries since https://github.com/cockroachdb/cockroach/pull/91630 so plumbing it into the kvserver will help paint a complete picture.

Note, the coordinator node of the job will always annotate its context with labels regardless of whether the coordinator node is being actively profiled or not. The performance impact of doing this is hypothesized to be negligible when compared to the long-running operation that the job will likely perform. Concretely, there are two ways we can go about propagating these labels to kvserver:

### Option 1:

Every request that wishes to have its labels propagated sets a field on the `BatchRequest.Header` containing the profiler labels that the sender's context is annotated with. `kvserver` then annotates the context that will serve the request if the node the request is being served on is actively being profiled. This ensures that we do not incur a performance hit when not actively profiling. Note, to get a complete picture, cluster-wide profiling would have to be enabled so as to propagate labels to requests across all nodes.

This option is self-contained, and can probably be backported to stable release branches as well.

### Option 2:

As prototyped in https://github.com/cockroachdb/cockroach/pull/60508 we set up gRPC interceptors that propagate the labels across RPC boundaries. This option would be more generalized and allows non-job SQL queries to also benefit from the propagation of labels. This option would probably be harder to backport.

Epic: CRDB-8964

Jira issue: CRDB-26281

### PR DESCRIPTION
**Pull request by adityamaru**
This change teaches DistSender to populate a BatchRequest's header with the pprof labels set on the sender's context. If the node processing the request on the server side has CPU profiling with labels enabled, then the labels stored in the BatchRequest will be applied to the root context of the goroutine executing the request on the server. Doing so will ensure that all server-side CPU samples of the root goroutine and all its spawned goroutine will be labeled correctly.

Propagating these labels across RPC boundaries is useful to correlate server side samples with the sender. For example, in a CPU profile generated with this change, we will be able to identify which backup job sent an ExportRequest that is causing a CPU hotspot on a remote node.

Fixes: #100166
Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
I wanted to get some feedback on this approach and whether we need a benchmark on how expensive it is for us to check for pprof labels on each sent BatchRequest. I tried this patch out on a roachprod cluster where the majority of ExportRequests were sent from n7,8,9 but were processed on n1,2,3 and it seems to work as expected. This is a profile on remote n1 while the coordinator of the job is on n9:

```
❯ go tool pprof profile.pb.gz
File: cockroach
Type: cpu
Time: Apr 12, 2023 at 12:52pm (EDT)
Duration: 5.14s, Total samples = 1.79s (34.85%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) tags
 distsql.gateway: Total 610.0ms
                  610.0ms (  100%): 8

 f: Total 610.0ms
    610.0ms (  100%): 73b7a58d

 job: Total 610.0ms
      610.0ms (  100%): BACKUP id=856056997655871496

 n: Total 620.0ms
    610.0ms (98.39%): 9
     10.0ms ( 1.61%): 3

 pebble: Total 240.0ms
         180.0ms (75.00%): compact
          60.0ms (25.00%): wal-sync

 r: Total 10.0ms
    10.0ms (  100%): 58

 range_str: Total 610.0ms
            260.0ms (42.62%): 579/6:/Table/108/1/167{3/5/…-5/3/…}
            240.0ms (39.34%): 1352/2:/Table/108/1/16{84/8/…-96/3/…}
            110.0ms (18.03%): 1348/2:/Table/108/1/167{5/3/…-9/2/…}

 remote_node_id: Total 50.0ms
                 40.0ms (80.00%): 7
                 10.0ms (20.00%): 9

 s: Total 10.0ms
    10.0ms (  100%): 3
```

### Profile with `-tagfocus` set to our BACKUP job.

Screenshot 2023-04-12 at 9 05 24 PM

Screenshot 2023-04-12 at 9 07 08 PM

File: pkg/server/node.go [hunk 1]
```diff
@@ -1250,6 +1250,16 @@ func (n *Node) Batch(ctx context.Context, args *kvpb.BatchRequest) (*kvpb.BatchR
 		ctx = logtags.AddTag(ctx, "tenant", tenantID.String())
 	}
 
+	// If the node is collecting a CPU profile with labels, and the sender has set
+	// pprof labels in the BatchRequest, then we apply them to the context that is
+	// going to execute the BatchRequest. These labels will help correlate server
+	// side CPU profile samples to the sender.
+	if n.execCfg.Settings.CPUProfileType() == cluster.CPUProfileWithLabels && len(args.ProfileLabels) != 0 {
```
Comment (dt) [inline pkg/server/node.go:1257]: nit: might be marginally better to flip the order of the two conditions since the len check is marginally cheaper?

**Comment by adityamaru** [comment]
@tbg I'm curious to get your take on this approach. We wanted something that we could potentially backport to older releases and thought that this change gave us enough coverage to begin with. The alternative would be writing grpc interceptors to get _complete_ coverage but that will probably be harder to reason about for a backport. Are there any benchmarks or optimizations that you would like to see as part of this change?

File: pkg/server/node.go (hunk 1 above)
Comment (adityamaru) [reply pkg/server/node.go]: done

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101404)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @miretskiy, and @tbg)

___
*[`pkg/server/node.go` line 1251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101404#-NTUtpcB-wpteBNNwwio:-NTUtpcB-wpteBNNwwip:bxme19i) ([raw file](https://github.com/cockroachdb/cockroach/blob/307b40ddbe6a13b46baddce9efdcee2210881b5a/pkg/server/node.go#L1251)):*
> ```Go
> 	} else {
> 		// We had this tag before the ResetAndAnnotateCtx() call above.
> 		ctx = logtags.AddTag(ctx, "tenant", tenantID.String())
> ```

This should really be `AddTag(ctx, "tenant", tenantID)` so it doesn't become redactable. Feel free to add a commit to do that.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101404)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, @miretskiy, and @tbg)

___
*[`pkg/server/node.go` line 1251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101404#-NTUtpcB-wpteBNNwwio:-NTVBDZB3dIpbrEOtu4C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/307b40ddbe6a13b46baddce9efdcee2210881b5a/pkg/server/node.go#L1251)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This should really be `AddTag(ctx, "tenant", tenantID)` so it doesn't become redactable. Feel free to add a commit to do that.
</blockquote></details>

Done.

**Review by dt — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!

bors r=dt

**Comment by adityamaru** [comment]
blathers backport 23.1 22.2

**Comment by adityamaru** [comment]
I will let this bake on master for a couple of weeks before merging the backports

**Pull request merged**